File: sorts/sliding_window_maximum_deque.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if not nums or k == 0:
            return []
        
        result = []
        dq = deque()

        for i, nums in enumerate(len(nums)):
            if dq and dq[0] <= i - k:
                dq.popleft()

            while dq and nums[dq[-1]] < num:
                dq.pop
            dq.append(i)

            if i >= k - 1:
                result.append(nums[dq[0]])

        return result

File: searches/sliding_window_maximum_deque.rb
def max_sliding_window(nums, k)
    return [] if nums.empty? || k == 0
    result = []
    deque = []

    nums.each_with_index do |num, i|
        deque.shift if !deque.empty? && deque[0] <= i - k

        deque.pop while !deque.empty? && nums[deque[-1]] < num
    
        deque << i

        result << nums[deque[0]] if i >= k - 1
    end
    
    result
end
